File: zig/src/indicators/core/component_triple_mnemonic.zig
const bar_component = @import("bar_component");
const quote_component = @import("quote_component");
const trade_component = @import("trade_component");

/// Builds a mnemonic suffix from bar, quote and trade components.
///
/// A component equal to its default is omitted. For example, if the bar
/// component is Median (non-default), the result is ", hl/2". If the bar
/// component is Close (default), it is omitted.
pub fn componentTripleMnemonic(
    buf: []u8,
    bc: bar_component.BarComponent,
    qc: quote_component.QuoteComponent,
    tc: trade_component.TradeComponent,
) []const u8 {
    var pos: usize = 0;

    if (bc != bar_component.default_bar_component) {
        const m = bar_component.componentMnemonic(bc);
        @memcpy(buf[pos .. pos + 2], ", ");
        pos += 2;
        @memcpy(buf[pos .. pos + m.len], m);
        pos += m.len;
    }

    if (qc != quote_component.default_quote_component) {
        const m = quote_component.componentMnemonic(qc);
        @memcpy(buf[pos .. pos + 2], ", ");
        pos += 2;
        @memcpy(buf[pos .. pos + m.len], m);
        pos += m.len;
    }

    if (tc != trade_component.default_trade_component) {
        const m = trade_component.componentMnemonic(tc);
        @memcpy(buf[pos .. pos + 2], ", ");
        pos += 2;
        @memcpy(buf[pos .. pos + m.len], m);
        pos += m.len;
    }

    return buf[0..pos];
}

File: rust/src/indicators/core/component_triple_mnemonic.rs
use crate::entities::bar_component::{component_mnemonic as bar_mnemonic, BarComponent, DEFAULT_BAR_COMPONENT};
use crate::entities::quote_component::{component_mnemonic as quote_mnemonic, QuoteComponent, DEFAULT_QUOTE_COMPONENT};
use crate::entities::trade_component::{component_mnemonic as trade_mnemonic, TradeComponent, DEFAULT_TRADE_COMPONENT};

/// Builds a mnemonic suffix from bar, quote and trade components.
///
/// A component equal to its default is omitted. For example, if the bar
/// component is Median (non-default), the result is `", hl/2"`. If the bar
/// component is Close (default), it is omitted.
pub fn component_triple_mnemonic(
    bc: BarComponent,
    qc: QuoteComponent,
    tc: TradeComponent,
) -> String {
    let mut s = String::new();

    if bc != DEFAULT_BAR_COMPONENT {
        s.push_str(", ");
        s.push_str(bar_mnemonic(bc));
    }

    if qc != DEFAULT_QUOTE_COMPONENT {
        s.push_str(", ");
        s.push_str(quote_mnemonic(qc));
    }

    if tc != DEFAULT_TRADE_COMPONENT {
        s.push_str(", ");
        s.push_str(trade_mnemonic(tc));
    }

    s
}
